File: Lab5(maybe_final)/address_book.py
from typing import List, Dict
from address import Address
from private_house import PrivateHouse
from appartment import Appartment
import sys  # для Exception
from exceptions import NotAddressFormatError, log_exception, NotNumberError, WrongNumberError


class AddressBook:
    def __init__(self):
        self.addresses: List[Address] = []      # список адрес
        self.selected_indices: List[int] = []   # індекси вибраних адрес для розсилки

    def add_address(self, address: Address):
        try:
            self.addresses.append(address)
        except NotAddressFormatError as e:
            log_exception(e)  # Логування помилки
            print(f"Caught exception: {e}", file=sys.stderr)
            raise

            
    def analyze_addresses(self):
        private_house_count = 0
        appartment_count = 0
        appartment_distribution: Dict[int, int] = {}

        if not self.selected_indices:
            print("There are no addresses selected for analysis.")
            return

        # Обробляємо вибрані адреси
        for index in self.selected_indices:
            address = self.addresses[index]
            if address.get_type() == "Private":
                private_house_count += 1
            elif address.get_type() == "Appartment":
                appartment_count += 1
                building_number = address.building_number
                if building_number in appartment_distribution:
                    appartment_distribution[building_number] += 1
                else:
                    appartment_distribution[building_number] = 1

        building_count = len(appartment_distribution)
        print("\nAnalysis Results:")
        print(f"Number of private houses: {private_house_count}")
        print(f"Number of appartments: {appartment_count}")
        print(f"Number of unique buildings: {building_count}")
        if building_count > 0:
            average_appartments = appartment_count / building_count
            print(f"Average number of appartments per building: {average_appartments:.2f}")
            for building, count in appartment_distribution.items():
                print(f"Building {building}: {count} appartments")

    def display_addresses(self):
        for idx, address in enumerate(self.addresses, start=1):
            print(f"{idx}: {address.to_string()}")

    def select_addresses_for_mailing(self):
        # Дозволяє користувачу обрати адреси для розсилки
        self.display_addresses()
        input_str = input("Enter the numbers of the addresses (separated by spaces) for mailing:  ")
        selected_numbers = input_str.split()

        self.selected_indices.clear()
        for num_str in selected_numbers:
            try:
                # Перевірка, чи введене значення є цілим числом
                if not num_str.isdigit():
                    raise NotNumberError(f"Input contains non-integer values - {num_str}. Please enter valid integer numbers.")

                number = int(num_str) - 1
                # Перевірка, чи введений індекс знаходиться в межах доступних адрес
                if not (0 <= number < len(self.addresses)):
                    raise WrongNumberError(f"Number {num_str} is out of range. Please select valid addresses.")

                self.selected_indices.append(number)

            except NotNumberError as e:
                log_exception(e)  # Логування помилки
                print(f"Caught exception: {e}", file=sys.stderr)

            except WrongNumberError as e:
                log_exception(e)  # Логування помилки
                print(f"Caught exception: {e}", file=sys.stderr)

            except Exception as e:
                log_exception(e)  # Логування помилки
                print(f"An unexpected error occurred: {e}", file=sys.stderr)
            
            # Перевірка наявності дублікатів у списку
        if len(self.selected_indices) != len(set(self.selected_indices)):
            print("Warning: There are duplicate indices in your selection.")
            user_choice = input("Enter 'R' to re-enter values or 'A' to automatically remove duplicates: ").strip().upper()

            if user_choice == 'R':
                print("Please enter the addresses again.")
                self.select_addresses_for_mailing()  # Рекурсивний виклик для повторного введення
            elif user_choice == 'A':
                # Видалення дублікатів за допомогою перетворення у множину
                self.selected_indices = list(set(self.selected_indices))
                print("Duplicates have been automatically removed.")
            else:
                print("Invalid choice. No action taken.")

        if self.selected_indices:
            print("You have selected addresses with numbers:", ", ".join(str(i + 1) for i in self.selected_indices))
        else:
            print("No addresses have been selected.")

File: Lab5_maybeFinal/addressBook.cpp
#include "addressBook.h"
#include "Exceptions.h"
#include <fstream>
#include <iostream>
#include <string>
#include <map>
#include <sstream>
#include <iomanip>
#include <set>
#include <algorithm>

AddressBook::AddressBook() {}

void AddressBook::AddAddress(shared_ptr<Address> address) {
    try {
        if (!address) {
            throw NotAddressFormatError("Invalid address format: Address is null");
        }
        addresses.push_back(address);
    }
    catch (const NotAddressFormatError& e) {
        // removes this fucking coments
        log_exception(e);  // Логування помилки
        cerr << "Caught exception: " << e.what() << endl;  // Виведення повідомлення про помилку
    }
    catch (const exception& e) {
        log_exception(e);  // Логування інших помилок
        cerr << "An unexpected error occurred: " << e.what() << endl;
    }
}

void AddressBook::AnalyzeAddresses() {
    // потрібно напрям прописати модифікатори доступу, класів це також стосується
    int privateHouseCount = 0;
    int appartmentCount = 0;
    std::map<int, int> appartmentDistribution;

    if (selectedIndices.empty()) {
        std::cout << "There are no addresses selected for analysis." << std::endl;
        return;
    }

    for (int index : selectedIndices) {
        if (index >= 0 && index < addresses.size()) {
            const auto& address = addresses[index];
            // хз чи яке таке в плюсах але в шарпах були порівнянян набагто кращі ніж по строці. Можливо є шось типу is або typeof 
            if (address->GetType() == "Private") {
                privateHouseCount++;
            }
            else if (address->GetType() == "Appartment") {
                appartmentCount++;
                //мені вс підсвічує що dynamic_cast писати не обовязково
                int buildingNumber = dynamic_cast<Appartment*>(address.get())->GetBuildingNumber();
                appartmentDistribution[buildingNumber]++;
            }
        }
        else {
            std::cerr << "Warning: Invalid index " << index << " (out of range)" << std::endl;
        }
    }

    int buildingCount = appartmentDistribution.size();
    std::cout << "\nAnalysis Results:" << std::endl;
    std::cout << "Number of private houses: " << privateHouseCount << std::endl;
    std::cout << "Number of appartments: " << appartmentCount << std::endl;
    std::cout << "Number of unique buildings: " << buildingCount << std::endl;

    if (buildingCount > 0) {
        //аналогічно static_cast
        double averageAppartments = static_cast<double>(appartmentCount) / buildingCount;
        std::cout << "Average number of appartments per building: " << averageAppartments << std::endl;
        for (const auto& pair : appartmentDistribution) {
            std::cout << "Building " << pair.first << ": " << pair.second << " appartments" << std::endl;
        }
    }
    else if (appartmentCount > 0) {
        std::cout << "Warning: Apartments found but building numbers could not be determined." << std::endl;
    }
}

void AddressBook::DisplayAddresses() const {
    for (size_t i = 0; i < addresses.size(); i++) {
        //в плюсах не можна в строку додатвати текст типу "Hello {value} Word"?, якщо можна то краще робити так
        std::cout << i + 1 << ": " << addresses[i]->toString() << std::endl;
    }
}


void AddressBook::SelectAddressesForMailing() {

    DisplayAddresses();
    string inputStr;
    cout << "Enter the numbers of the addresses (separated by spaces) for mailing: ";
    getline(cin, inputStr);

    selectedIndices.clear();
    istringstream iss(inputStr);
    string numStr;
    set<int> duplicatesChecker;

    while (iss >> numStr) {
        try {
            //прибрати коментарі
            // !all_of(numStr.begin(), numStr.end(), ::isdigit) оцю фігню можна в окремий метод винести щоб було зрозуміло що тут відбувається
            // Перевірка, чи є введене значення числом
            if (!all_of(numStr.begin(), numStr.end(), ::isdigit)) {
                throw NotNumberError("Input contains non-integer values - " + numStr + ". Please enter valid integer numbers.");
            }

            int number = stoi(numStr) - 1;
            //зауваження аналогічні
            // Перевірка, чи введений індекс у межах допустимих значень
            if (number < 0 || number >= static_cast<int>(addresses.size())) {
                throw WrongNumberError("Number " + numStr + " is out of range. Please select valid addresses.");
            }

            selectedIndices.push_back(number);
            //тут пробіл не потрібний
        }
        catch (const NotNumberError& e) {
            log_exception(e);
            cerr << "Caught exception: " << e.what() << endl;

        }
        catch (const WrongNumberError& e) {
            log_exception(e);
            cerr << "Caught exception: " << e.what() << endl;

        }
        catch (const exception& e) {
            log_exception(e);
            cerr << "An unexpected error occurred: " << e.what() << endl;
        }
    }

    //це все ще одна функція? фігня повна - розбити

    // Перевірка дублікатів
    duplicatesChecker.insert(selectedIndices.begin(), selectedIndices.end());
    if (duplicatesChecker.size() != selectedIndices.size()) {
        cout << "Warning: There are duplicate indices in your selection." << endl;
        cout << "Enter 'R' to re-enter values or 'A' to automatically remove duplicates: ";
        char choice;
        cin >> choice;
        choice = toupper(choice);
        //що таке R і A - це якийсь вибір але мені просто по букві не зрозуміло - винести в enum
        if (choice == 'R') {
            cin.ignore(numeric_limits<streamsize>::max(), '\n');
            cout << "Please enter the addresses again." << endl;
            //коментар
            SelectAddressesForMailing();  // Рекурсивний виклик
            return;
        }
        else if (choice == 'A') {
            cin.ignore(numeric_limits<streamsize>::max(), '\n');
            selectedIndices.assign(duplicatesChecker.begin(), duplicatesChecker.end());
            cout << "Duplicates have been automatically removed." << endl;
        }
        else {
            cout << "Invalid choice. No action taken." << endl;
        }
    }

    // Підсумковий результат
    if (!selectedIndices.empty()) {
        cout << "You have selected addresses with numbers: ";
        for (int index : selectedIndices) {
            cout << index + 1 << " ";
        }
        cout << endl;
    }
    else {
        cout << "No addresses have been selected." << endl;
    }
}

//додати усюди явно модифікатори доступу
void AddressBook::ClearAddresses() {
    addresses.clear();
}

// Геттер для `addresses` - повертає список адрес
const std::vector<std::shared_ptr<Address>>& AddressBook::GetAddresses() const {
    return addresses;
}

// Геттер для `selectedIndices` - повертає список обраних індексів
const std::vector<int>& AddressBook::GetSelectedIndices() const {
    return selectedIndices;
}
